File: tdms_read.py
from nptdms import TdmsFile
import os

def print_tdms_structure(tdms_file_path):
    try:
        tdms_file = TdmsFile.read(tdms_file_path)
        print(f"Inspecting TDMS File: {tdms_file_path}\n")
        
        print("Groups and Channels:")
        for group in tdms_file.groups():
            print(f"\n🔹 Group: {group.name}")
            for channel in group.channels():
                print(f"    Channel: {channel.name}")
                data = channel[:]
                print(f"        - Data Sample (first 5 points): {data[:5]}")
                
    except Exception as e:
        print(f"An error occurred while reading the TDMS file: {e}")

if __name__ == "__main__":
    tdms_file_path ='/current,temp/0Nm_BPFI_03.tdms'
    
    if os.path.exists(tdms_file_path):
        print_tdms_structure(tdms_file_path)
    else:
        print(f"TDMS file not found at {tdms_file_path}. Please check the path and try again.")

from nptdms import TdmsFile
import pandas as pd
import numpy as np
import os

# Define the sampling rate (samples per second)
SAMPLING_RATE = 25600  # 25.6 kHz

def construct_time_stamp(num_samples, sampling_rate):
    increment = 1.0 / sampling_rate
    return np.round(np.arange(num_samples) * increment, decimals=6)
    
def extract_channels(group, channel_mapping):
    data = {}
    for tdms_channel, csv_column in channel_mapping.items():
        try:
            channel = group[tdms_channel]
            data[csv_column] = channel[:]
        except KeyError:
            print(f"Channel '{tdms_channel}' not found in group '{group.name}'. Skipping this channel.")
    return data

def convert_tdms_to_csv(tdms_directory, csv_directory, group_name, channel_mapping, sampling_rate):
    # Ensure the CSV directory exists
    os.makedirs(csv_directory, exist_ok=True)

    # Iterate through all TDMS files in the directory
    for filename in os.listdir(tdms_directory):
        if filename.lower().endswith('.tdms'):
            tdms_path = os.path.join(tdms_directory, filename)
            print(f"Processing TDMS File: {tdms_path}")
            try:
                tdms_file = TdmsFile.read(tdms_path)
                
                # Group Check
                group_found = False
                for group in tdms_file.groups():
                    if group.name == group_name:
                        selected_group = group
                        group_found = True
                        break

                if not group_found:
                    print(f"Group '{group_name}' not found in {filename}. Skipping.\n")
                    continue

                # Extract Channels
                extracted_data = extract_channels(selected_group, channel_mapping)

                if not extracted_data:
                    print(f"No channels extracted from {filename}. Skipping.\n")
                    continue

                # Determine the number of samples from one of the channels
                num_samples = len(next(iter(extracted_data.values())))

                # Construct Time Stamp
                time_stamp = construct_time_stamp(num_samples, sampling_rate)

                # Create DataFrame
                df = pd.DataFrame({
                    'Time Stamp (s)': time_stamp,
                    'Temperature_housing_A (°C)': extracted_data.get('Temperature_housing_A', pd.NA),
                    'Temperature_housing_B (°C)': extracted_data.get('Temperature_housing_B', pd.NA),
                    'U-phase (A)': extracted_data.get('U-phase', pd.NA),
                    'V-phase (A)': extracted_data.get('V-phase', pd.NA),
                    'W-phase (A)': extracted_data.get('W-phase', pd.NA)
                })

                # Define the CSV file path
                csv_filename = os.path.splitext(filename)[0] + '.csv'
                csv_path = os.path.join(csv_directory, csv_filename)

                # Save to CSV
                df.to_csv(csv_path, index=False)
                print(f"Saved CSV to {csv_path}\n")

            except Exception as e:
                print(f"An error occurred while processing {filename}: {e}\n")
                continue


tdms_directory = './current,temp'
csv_directory = './current_temp'

# Group name
group_name = 'Log'

# Define channel mapping: TDMS channel names to desired CSV column names
channel_mapping = {
    'cDAQ9185-1F486B5Mod1/ai0': 'Temperature_housing_A',
    'cDAQ9185-1F486B5Mod1/ai1': 'Temperature_housing_B',
    'cDAQ9185-1F486B5Mod2/ai0': 'U-phase',
    'cDAQ9185-1F486B5Mod2/ai2': 'V-phase',
    'cDAQ9185-1F486B5Mod2/ai3': 'W-phase'
}

# Sampling rate (Hz) - 25.6 kHz
SAMPLING_RATE = 25600  # 25.6 kHz

# Convert TDMS files to CSV
convert_tdms_to_csv(tdms_directory, csv_directory, group_name, channel_mapping, SAMPLING_RATE)

File: tdms_read.m
% convert_current_temp_csvs.m
% Put this file in the project root (same folder that has 'current_temp' and other folders).
% Run: convert_current_temp_csvs
%
% Behavior:
% - Looks for 'current_temp' folder in pwd
% - Creates 'current_temp_csv' folder in pwd (if needed)
% - Reads each CSV in current_temp, handles TDMS-exported headers like /'Log'/'cDAQ...
% - Detects if Time Stamp column exists; if not, creates Time Stamp = (0:N-1)' * dt
%   (default dt = 1.0 second). Change dt below if you know the true sampling interval.
% - Writes a standardized CSV with columns:
%   Time Stamp,Temperature_housing_A,Temperature_housing_B,U-phase,V-phase,W-phase,load,condition,severity

function convert_current_temp_csvs()
    root_dir = pwd;
    input_dir = fullfile(root_dir, 'current_temp');
    output_dir = fullfile(root_dir, 'current_temp_csv');

    if ~isfolder(input_dir)
        error('Input folder not found: %s', input_dir);
    end
    if ~isfolder(output_dir)
        mkdir(output_dir);
    end

    % Default sample interval (seconds) to use if file lacks Time Stamp
    dt_default = 1.0;  % change this to the true dt if known

    files = dir(fullfile(input_dir, '*.csv'));
    if isempty(files)
        fprintf('No CSV files found in %s\n', input_dir);
        return;
    end

    for k = 1:numel(files)
        fname = files(k).name;
        fpath = fullfile(input_dir, fname);
        try
            % read first line to inspect header
            fid = fopen(fpath, 'r');
            if fid < 0
                warning('Failed to open %s', fpath);
                continue;
            end
            firstLine = strtrim(fgetl(fid));
            fclose(fid);

            % Some TDMS->CSV exports put single-line channel header like:
            % /'Log'/'cDAQ9185-1F486B5Mod1/ai0',/'Log'/'cDAQ9185-1F486B5Mod1/ai1',...
            % We want to map them to the 5 expected columns (no timestamp) or detect a Time Stamp.
            % We'll try to read the CSV robustly with readtable.
            opts = detectImportOptions(fpath, 'NumHeaderLines', 0);
            % force reading as text first to avoid weird column names
            Traw = readtable(fpath, opts);

            % If the very first column name looks like a path (contains 'Log' or starts with "/'"),
            % we assume there is no Time Stamp and the file columns correspond to:
            %  [Temperature_housing_A, Temperature_housing_B, U-phase, V-phase, W-phase]
            colnames = Traw.Properties.VariableNames;

            % clean column names (remove quotes, slashes)
            cleaned = clean_colnames(colnames);

            hasTimeStamp = any(contains(lower(cleaned), 'time')) || any(strcmpi(cleaned, 'TimeStamp'));

            if hasTimeStamp
                % Ensure first column is a Time Stamp-like column; if not exactly named Time Stamp, rename.
                % Find the time column index
                timeIdx = find(contains(lower(cleaned), 'time'), 1, 'first');
                timeVec = Traw{:, timeIdx};
                % Remove the existing time column from Traw columns for reassignment below
                nonTimeIdx = setdiff(1:size(Traw,2), timeIdx);
                dataMatrix = Traw{:, nonTimeIdx};
                dataCols = cleaned(nonTimeIdx);
            else
                % No explicit time column; assume the table has 5 columns corresponding to:
                % Temperature_housing_A, Temperature_housing_B, U-phase, V-phase, W-phase
                dataMatrix = Traw{:,:};
                dataCols = cleaned;
                N = size(dataMatrix,1);
                timeVec = (0:(N-1))' * dt_default;
            end

            % Now map columns to target names. We expect 5 data columns.
            % If file has more or fewer columns, we'll try sensible mapping:
            % Prefer to pick columns by matching names, otherwise use order.
            targetNames = {'Temperature_housing_A', 'Temperature_housing_B', 'U-phase', 'V-phase', 'W-phase'};

            % Create table starting with Time Stamp
            T = table();
            T.("Time Stamp") = timeVec;

            % Map data columns into the 5 target columns
            M = size(dataMatrix,2);
            mapped = false(1,5);

            % 1) match by words in column names if possible
            for i = 1:M
                cname = dataCols{i};
                lname = lower(cname);
                if contains(lname, 'temp') && ~mapped(1)
                    T.("Temperature_housing_A") = dataMatrix(:, i);
                    mapped(1) = true;
                elseif contains(lname, 'housing') && contains(lname, 'b') && ~mapped(2)
                    T.("Temperature_housing_B") = dataMatrix(:, i);
                    mapped(2) = true;
                elseif contains(lname, 'u') && contains(lname, 'phase') && ~mapped(3)
                    T.("U-phase") = dataMatrix(:, i);
                    mapped(3) = true;
                elseif contains(lname, 'v') && contains(lname, 'phase') && ~mapped(4)
                    T.("V-phase") = dataMatrix(:, i);
                    mapped(4) = true;
                elseif contains(lname, 'w') && contains(lname, 'phase') && ~mapped(5)
                    T.("W-phase") = dataMatrix(:, i);
                    mapped(5) = true;
                end
            end

            % 2) fill remaining target columns by order from left to right
            j = 1; % index in dataMatrix
            for tname_i = 1:numel(targetNames)
                if ~isfield(T, targetNames{tname_i})
                    % find next unused column in dataMatrix
                    while j <= M && any(ismember(dataCols{j}, T.Properties.VariableNames))
                        j = j + 1;
                    end
                    if j <= M
                        T.(targetNames{tname_i}) = dataMatrix(:, j);
                        j = j + 1;
                    else
                        % if not enough columns, fill with NaNs
                        T.(targetNames{tname_i}) = NaN(height(T),1);
                    end
                end
            end

            % Add filename metadata columns (load, condition, severity)
            meta = parse_filename_metadata(fname);
            if ~isempty(meta)
                T.load = repmat({meta.load}, height(T), 1);
                T.condition = repmat({meta.condition}, height(T), 1);
                T.severity = repmat({meta.severity}, height(T), 1);
            end

            % Write out CSV to output_dir
            outname = fullfile(output_dir, [strip_extension(fname) '.csv']);
            writetable(T, outname);
            fprintf('Converted: %s -> %s\n', fname, outname);

        catch ME
            warning('Failed to process %s: %s', fname, ME.message);
        end
    end

    fprintf('Finished processing %d files.\n', numel(files));
end

%% ===== Helper functions =====
function s = strip_extension(fname)
    [~, s, ~] = fileparts(fname);
end

function cleaned = clean_colnames(colnames)
    % remove slashes, quotes, spaces and make simpler forms for matching
    cleaned = cell(size(colnames));
    for i=1:numel(colnames)
        c = colnames{i};
        % common TDMS CSV gives header like: /'Log'/'cDAQ9185-1F486B5Mod1/ai0'
        % remove slashes and single quotes
        c = strrep(c, '/', '');
        c = strrep(c, '''', '');
        c = strrep(c, '"', '');
        c = strrep(c, '?', '');
        c = strtrim(c);
        % replace non-alphanum with underscore
        c = regexprep(c, '[^a-zA-Z0-9]', '_');
        cleaned{i} = c;
    end
end

function meta = parse_filename_metadata(fname)
    % Parse filenames of form: aaaaNm_bbbb_cccc.*  ->  load, condition, severity
    meta = struct();
    name = strip_extension(fname);
    parts = strsplit(name, '_');
    if numel(parts) >= 3
        meta.load = parts{1};
        meta.condition = parts{2};
        meta.severity = parts{3};
    else
        meta = [];
    end
end
